🌱  controllers/topology: cleanup error handling and templateNamer

sbueringer opened the pull request:
Signed-off-by: Stefan Büringer buringerst@vmware.com

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

killianmuldoon commented on controllers/topology/util.go:
```diff
@@ -68,3 +69,8 @@ func refToUnstructured(ref *corev1.ObjectReference) *unstructured.Unstructured {
 }
+
+// objError creates an error based on the given err, msg and object.
+func objError(err error, msg string, object client.Object) error {
```
Is the idea to make this a pattern going forward? fmt.Errorf is used in a bunch of places across the code, and it's a well known function. This helper makes things easier to read in some ways, but if it's not an accepted pattern across the codebase it could make things harder to follow.

sbueringer replied:
The goal was to avoid repetition in the way we construct our errors in the topology controller. And also to make them more consistent. There were already a few cases where we built them differently, in some cases they were even wrong (we had `<gvk>/kind`)

I'll wait for further feedback. If we don't want it or someone has a better idea how to do it I'm happy to adjust the PR accordingly.

killianmuldoon commented on controllers/topology/reconcile_state.go:
```diff
@@ -83,26 +83,24 @@ func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, s *scope.
 		err = contract.ControlPlane().InfrastructureMachineTemplate().Set(s.Desired.ControlPlane.Object, refToUnstructured(cpInfraRef))
 		if err != nil {
-			return kerrors.NewAggregate([]error{errors.Wrapf(err, "failed to update the %s object", s.Desired.ControlPlane.Object.GetKind()), cleanup()})
+			return kerrors.NewAggregate([]error{objError(err, "failed to update the %s object", s.Desired.ControlPlane.Object), cleanup()})
```
This seems harder to follow to me than the original version - at least at first glance. With errors.Wrapf it's clear what's going in and out.

sbueringer commented on controllers/topology/reconcile_state.go:
```diff
@@ -78,7 +78,7 @@ func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, s *scope.
 		cpInfraRef, err := contract.ControlPlane().MachineTemplate().InfrastructureRef().Get(s.Desired.ControlPlane.Object)
 		if err != nil {
-			return errors.Wrapf(err, "failed to update the %s object,", s.Desired.ControlPlane.InfrastructureMachineTemplate.GetKind())
+			return errors.Wrapf(err, "failed to update %s", errObjectID(s.Desired.ControlPlane.InfrastructureMachineTemplate))
```
```suggestion
			return errors.Wrapf(err, "failed to update %s", KRef{Obj: s.Desired.ControlPlane.InfrastructureMachineTemplate})
```

@fabriziopandini The idea is to switch to KRef, right?
Just asking because previously our format was:
```go
fmt.Sprintf("%s, %s/%s", object.GetObjectKind().GroupVersionKind().GroupVersion(), object.GetObjectKind().GroupVersionKind().Kind, object.GetName()
```
And KRef does not include the group version:
```go
fmt.Sprintf("%s/%s", ref.Obj.GetObjectKind().GroupVersionKind().Kind, ref.Obj.GetName())
```

But I think it's fine to drop the group version as it will make the error message more concise and I think we don't really need the group version to identify the object.

fabriziopandini replied:
Agreed. Let's KRef

sbueringer commented:
Rebased onto the main branch and now the error's are simply using KRef (like our logs)

killianmuldoon commented on controllers/topology/util.go:
```diff
@@ -114,18 +114,30 @@ func (l *topologyReconcileLogger) Into(ctx context.Context) (context.Context, lo
-// KRef return a reference to a Kubernetes object in the same format used by kubectl commands (kind/name).
-type KRef struct {
+// KObj return a reference to a Kubernetes object in the same format used by kubectl commands (kind/name).
+type KObj struct {
```
Query - why use this struct over the klog.KObj / Kobjs functions from klog? Kubernetes seems to prefer to use those functions directly. It would be good to document here why this was chosen as an alternative.

sbueringer replied:
Not 100% sure as it was introduced in @fabriziopandini 's error PR. Afaik because the klog struct uses the `namespace/name` format while we need `kind/name`

killianmuldoon replied:
You're right - I missed it there. It would be good to get the reasoning documented in a comment as it looks strange for anyone used to the k/k structured klogs implementation. This could also be changed to a function to make it easier to read and write (like the klog implementation), but I don't think it's a blocker.

fabriziopandini replied:
klog.KRef reports name and namespace; here we are reporting Kind and name (and assuming the namespace is always the same); also we need to stardardize logs/error around OwnerReferences

sbueringer replied:
Hm, what about the point about making this a func. I have no idea why it's a struct in klog and what we loose when I replace the struct by a simple func.

sbueringer replied:
Note: would be good if can get consensus and merge this PR relatively soon, because we have multiple big PRs inflight and it would be great if we could align to this already.

killianmuldoon replied:
I think it's fine to leave this as a struct for now and we can worry about making it a function later on.

sbueringer replied:
@killianmuldoon I added a comment to KObj. Afaik there is no comparable struct in klog to `KRef`.

killianmuldoon replied:
Great - I think that's good enough for context around this if we look to expand this across the code base in future.

sbueringer replied:
Yup, agree

killianmuldoon reviewed (commented):
This is great for standardizing our approach - just one question to be answered from my PoV.

sbueringer commented:
@vincepri Would be good if you can also take a look, if you have some time.

killianmuldoon commented:
/lgtm

sbueringer commented:
/retest
the usual flake

fabriziopandini commented:
/approve

vincepri commented on controllers/topology/util.go:
```diff
@@ -114,18 +114,31 @@ func (l *topologyReconcileLogger) Into(ctx context.Context) (context.Context, lo
-type KRef struct {
+// KObj return a reference to a Kubernetes object in the same format used by kubectl commands (kind/name).
+// Note: We're intentionally not using klog.KObj as we want the kind/name format instead of namespace/name.
+type KObj struct {
```
Should this be exported?

sbueringer replied:
I think we should move it into an internal package. Otherwise it cannot be used by other packages in topology/* (either because it's not exported or because of circular dependencies)

fabriziopandini replied:
I'm +1 to keep this internal at this stage (no preferences if this should be in separated package).

sbueringer replied:
Done, moved to internal/log

vincepri commented on controllers/topology/util.go:
```diff
@@ -114,18 +114,31 @@ func (l *topologyReconcileLogger) Into(ctx context.Context) (context.Context, lo
 	return fmt.Sprintf("%s/%s", ref.Obj.GetObjectKind().GroupVersionKind().Kind, ref.Obj.GetName())
 }
 
+// KRef return a reference to a Kubernetes object in the same format used by kubectl commands (kind/name).
+type KRef struct {
```
Same as above, could we have an internal package instead of making this an API?

sbueringer replied:
Done

sbueringer commented:
Moved our log utils in a separate package and rebased onto current main. PTAL @vincepri @fabriziopandini 

Note: the PR is in theory already approved but I'll wait for lgtm from @fabriziopandini and @vincepri before I remove the hold

fabriziopandini commented:
/lgtm
/approve

sbueringer commented:
@fabriziopandini Should I/we remove the hold?

fabriziopandini commented:
/hold cancel
